sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,821 @@
+# Pairwise Distances Reductions
+# =============================
+#
+#    Author: Julien Jerphanion <git@jjerphan.xyz>
+#
+#
+# The routines defined here are used in various algorithms performing
+# the same structure of operations on distances between vectors
+# of a datasets pair (X, Y).
+#
+# Importantly, the core of the computation is chunked to make sure that the pairwise
+# distance chunk matrices stay in CPU cache before applying the final reduction step.
+# Furthermore, the chunking strategy is also used to leverage OpenMP-based parallelism
+# (using Cython prange loops) which gives another multiplicative speed-up in
+# favorable cases on many-core machines.
+cimport numpy as np
+import numpy as np
+import warnings
+import scipy.sparse
+
+from .. import get_config
+from libc.stdlib cimport free, malloc
+from libc.float cimport DBL_MAX
+from cython cimport final
+from cython.parallel cimport parallel, prange
+
+from ._dist_metrics cimport DatasetsPair, DenseDenseDatasetsPair
+from ..utils._cython_blas cimport (
+  BLAS_Order,
+  BLAS_Trans,
+  ColMajor,
+  NoTrans,
+  RowMajor,
+  Trans,
+  _dot,
+  _gemm,
+)
+from ..utils._heap cimport simultaneous_sort, heap_push
+from ..utils._openmp_helpers cimport _openmp_thread_num
+from ..utils._typedefs cimport ITYPE_t, DTYPE_t, DITYPE_t
+
+from numbers import Integral, Real
+from typing import List
+from scipy.sparse import issparse
+from ._dist_metrics import BOOL_METRICS, METRIC_MAPPING
+from ..utils import check_scalar, _in_unstable_openblas_configuration
+from ..utils.fixes import threadpool_limits
+from ..utils._openmp_helpers import _openmp_effective_n_threads
+from ..utils._typedefs import ITYPE, DTYPE
+
+np.import_array()
+
+cpdef DTYPE_t[::1] _sqeuclidean_row_norms(
+    const DTYPE_t[:, ::1] X,
+    ITYPE_t num_threads,
+):
+    """Compute the squared euclidean norm of the rows of X in parallel.
+
+    This is faster than using np.einsum("ij, ij->i") even when using a single thread.
+    """
+    cdef:
+        # Casting for X to remove the const qualifier is needed because APIs
+        # exposed via scipy.linalg.cython_blas aren't reflecting the arguments'
+        # const qualifier.
+        DTYPE_t * X_ptr = <DTYPE_t *> &X[0, 0]
+        ITYPE_t idx = 0
+        ITYPE_t n = X.shape[0]
+        ITYPE_t d = X.shape[1]
+        DTYPE_t[::1] row_norms = np.empty(n, dtype=DTYPE)
+
+    for idx in prange(n, schedule='static', nogil=True, num_threads=num_threads):
+        row_norms[idx] = _dot(d, X_ptr + idx * d, 1, X_ptr + idx * d, 1)
```

Comment:
I have not: BLAS LEVEL1 operations (used under the hood by `_dot`) have been included to avoid their loopy counterparts. See page 3 of https://dl.acm.org/doi/10.1145/355841.355847 for reference.

Reply:
As info: There is a culture that 👍 in a review comment means that the raised issue has been addressed.